### PR TITLE
VaultActions and LeverActions for Sense Finance

### PR DESCRIPTION
**Pull request by lacoop6tu**
WIP
- add initial actions integration and tests

### CODE REVIEW

File: src/test/utils/VaultSPT.sol [hunk 1]
```diff
@@ -0,0 +1,161 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.8.4;
+
```
Comment (SolcAndMe) [inline src/test/utils/VaultSPT.sol:3]: remove empty line

File: src/test/utils/VaultSPT.sol [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+import {SafeERC20} from "openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import {ICodex} from "fiat/interfaces/ICodex.sol";
+import {ICollybus} from "fiat/interfaces/ICollybus.sol";
+import {IVault} from "fiat/interfaces/IVault.sol";
+import {Guarded} from "fiat/utils/Guarded.sol";
+import {WAD, toInt256, wmul, wdiv} from "fiat/utils/Math.sol";
+
+/// @title VaultSPT
+
```
Comment (SolcAndMe) [inline src/test/utils/VaultSPT.sol:16]: remove empty line

File: src/test/utils/VaultSPT.sol [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+contract VaultSPT is Guarded, IVault {
```
Comment (SolcAndMe) [inline src/test/utils/VaultSPT.sol:17]: remove contract and import it from vaults

File: src/vault/VaultSPTActions.sol [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+
+import {ICodex} from "fiat/interfaces/ICodex.sol";
+import {IVault} from "fiat/interfaces/IVault.sol";
+import {IMoneta} from "fiat/interfaces/IMoneta.sol";
+import {IFIAT} from "fiat/interfaces/IFIAT.sol";
+import {WAD, toInt256, wmul, wdiv, sub} from "fiat/utils/Math.sol";
+
+import {Vault20Actions} from "./Vault20Actions.sol";
+
+
```
Comment (SolcAndMe) [inline src/vault/VaultSPTActions.sol:15]: align code style, remove empty lines throughout the contract ...

File: src/vault/VaultSPTActions.sol [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+        int256 deltaCollateral = -toInt256(wdiv(pTokenAmount, IVault(vault).tokenScale()));
+     
+        // withdraw pToken from the position
+        // swapParams.assetIn is the sP token
+        modifyCollateralAndDebt(vault, token, 0, position, address(this), creditor, deltaCollateral, deltaNormalDebt);
+
+        // Approve divider for pToken
+        IERC20(swapParams.assetIn).approve(address(divider),pTokenAmount);
+         // Redeem for target token
+        uint256 tBal = divider.redeem(swapParams.adapter,swapParams.maturity,pTokenAmount);
```
Comment (SolcAndMe) [inline src/vault/VaultSPTActions.sol:223]: naming not aligned with rest of contracts

**Review by SolcAndMe — CHANGES_REQUESTED** [review]

**Review by SolcAndMe — APPROVED** [review]

**Review by amarcu — COMMENTED** [review]
Looks good, added only a few minor typos.

File: src/lever/LeverSPTActions.sol [hunk 6]
```diff
@@ -0,0 +1,573 @@
...
+    }
+
+    /// ======== Position Management ======== ///
+
+    /// @notice Increases the leverage factor of a position by flash minting `addDebt` amount of FIAT
+    /// and selling it on top of the `underlierAmount` the `collateralizer` provided for more pTokens.
+    /// @param vault Address of the Vault
+    /// @param position Address of the position's owner
+    /// @param collateralizer Address of the account who puts up the up front amount of underliers
+    /// @param upfrontUnderliers Amount of underliers the `collateralizer` puts up up front [underlierScale]
```
[the remark this answers is not in the document]
Comment (SolcAndMe) [reply src/lever/LeverSPTActions.sol:188]: no this correct. upfront should be together though.

**Comment by SolcAndMe** [comment]
same typos are prob. also in LeverEPTActions

**Comment by SolcAndMe** [comment]
@lacoop6tu

File: src/lever/LeverSPTActions.sol [hunk 7]
```diff
@@ -0,0 +1,573 @@
...
+
+    /// @notice Decreases the leverage factor of a position by flash lending `subNormalDebt * rate` amount of FIAT
+    /// to decrease the outstanding debt (incl. interest), redeeming the withdrawn collateral (pToken) of the position
+    /// for underliers and selling a portion of the underliers for FIAT to repay the flash lent amount.
+    /// @param vault Address of the Vault
+    /// @param position Address of the position's owner
+    /// @param collateralizer Address of the account receives the excess collateral (redeemed underliers)
+    /// @param subPTokenAmount Amount of pToken to withdraw from the position [tokenScale]
+    /// @param subNormalDebt Amount of normalized debt of the position to reduce [wad]
+    /// @param fiatSwapParams Parameters of the underlier to FIAT swap
```
[the remark this answers is not in the document]
Comment (lacoop6tu) [reply src/lever/LeverSPTActions.sol:339]: added ty!

**Pull request merged**